core/src/main/java/org/pentaho/di/connections/utils/ConnectionUriParser.java

```diff
@@ -0,0 +1,229 @@
+ * This class handles special characters in the connection name section of the
+ * {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI by simply parsing
+ * the connection name as is. This class has limited functionality.
+ * <p/>
+ * Main motivation for this class is not being able to use directly {@link java.net.URI}
+ * due to special characters in the {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI.
+ * Future efforts should look into {@link org.apache.commons.vfs2.provider.UriParser#encode(String, char[])}.
+ *
+ * @see #SPECIAL_CHARACTERS_FULL_SET
+ */
+public class ConnectionUriParser {
+
+  /**
+   * Full set of special characters the connection name can be. Only excluding the character '/' which is a file path deliminator.
+   * <p/>
+   * Special characters in this context are characters that generally have to be encoded
+   * otherwise the use of {@link java.net.URI } will throw {@link java.net.URISyntaxException} when during
+   * object instantiation or subsequent method calls.
+   *
+   */
+  public static final String SPECIAL_CHARACTERS_FULL_SET = "!\"#$%&'()*+,-.:;<=>?@[\\] \t\n\r^_`{|}~";
+
+  /**
+   * Full set of alphanumeric characters the connection name can be.
+   */
+  public static final String ALPHANUMERIC_CHARACTERS_FULL_SET =
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+  /**
+   * Full set of accepted characters the connection name can be.
+   */
+  public static final String ACCEPTED_CHARACTERS_FULL_SET =
+      SPECIAL_CHARACTERS_FULL_SET + ALPHANUMERIC_CHARACTERS_FULL_SET;
+
+  /**
+   * Pattern to match a {@value org.pentaho.di.connections.vfs.provider.ConnectionFileProvider#SCHEME} URI
+   *  with scheme, connection name, and connection path.
+   * <p/> Some Examples:
+   * <ul>
+   *   <li>For PVFS URI: "pvfs://connectionName/someFolderA/SomeFolderB/someFile.txt"</li>
+   *      <ul>
+   *        <li>scehme: "pvfs"</li>
+   *        <li>connection Name: "connectionName"</li>
+   *        <li>pvfs path: "/someFolderA/SomeFolderB/someFile.txt"</li>
+   *
+   *      </ul>
+   * </ul>
+   * Regex should be encompass {@link org.pentaho.di.connections.vfs.provider.ConnectionFileSystem#DOMAIN_ROOT}
+   */
+  public static final Pattern CONNECTION_URI_WITH_CONNECTION_NAME_AND_PATH_PATTERN
+    = Pattern.compile(  "(\\w+)://([^/]+)(/.+)" );
+
+  /**
+   * Pattern to match a URI with scheme and connection name or first path segment.
+   * <p/> Some Examples:
+   * <ul>
+   *   <li>For PVFS URI: "pvfs://connectionName"</li>
+   *   <li>For VFS URI: "xyz://firstSegment"</li>
+   * </ul>
+   * Regex should be encompass {@link org.pentaho.di.connections.vfs.provider.ConnectionFileSystem#DOMAIN_ROOT}
+   */
+  public static final Pattern CONNECTION_URI_WITH_CONNECTION_NAME_PATTERN
+      = Pattern.compile(  "(\\w+)://([^/]+)/?" );
```

Comment:
Could `CONNECTION_URI_WITH_CONNECTION_NAME_AND_PATH_PATTERN` be changed to have the path part optional? Would it become too complex?

Reply:
I would suggest no code change on this part.

with the current code, I test 20+ strings for the various format and behave as expected.

1. Is there an edge case that I'm not supporting?
2. What objective and noticeable advantage would this gain?
3. Can you supply the code change you are asking for?

Reply:
No, unfortunately, I don't have time right now to supply the code change I am asking for.
As is, the code appears redundant and more complex than it needs to be.
Obviously, it's all relative. If the regex becomes hard to read/write, things change!

Reply:
I'm trying to triage all comments for this PR. I've stated my position and you have stated your position very well.

Is this a item that should prevent merging as ?

If so, we should  go offline and get a technical moderator to make a final decision.

Reply:
Resolving. See https://github.com/pentaho/pentaho-kettle/pull/9310#pullrequestreview-2065857215.